### DISCUSSION TITLE
Categorical subset sampling

### DISCUSSION THREAD

**Discussion by StochasticBoris** [optuna/optuna#3801] [Q&A]
Hi,

I was wondering if there is a canonical way to sample a subset of values from a categorical distribution.
If there are $N$ choices available, I would like to sample $M$ values from the choice set where $1<= M < N$ without replacement (sampling with replacement would be an interesting side topic). The order of the set does not matter.

An example would be as follows:

> Given $N$ columns in a dataset, select a subset $M$ columns on which to evaluate an objective function that is independent on their order.

Currently, I have attempted this by sampling each value in a loop:

```python
import random
import optuna

def objective(trial):
    all_columns = remaining_columns = {"column_1", "column_2", "column_3", "column_4", "column_5"}
    N = len(all_columns)
    # M = trial.suggest_int("M", 1, N -1) # May be constant or not
    M = 3

    sampled_columns = set()
    for i in range(M):        
        new_column = trial.suggest_categorical(f'column_{i}', remaining_columns)
        sampled_columns.add(new_column)
        remaining_columns.remove(new_column)

    score = random.uniform(0,1)
    return score 

study = optuna.create_study(direction="minimize")
study.optimize(objective, n_trials=10)
```

This has two problems:
1. The optimization fails outright on (usually) the second trial because of `ValueError: CategoricalDistribution does not support dynamic value space.`. This happens since the choices available in all loops following the first loop are dependent on the sampled values of the prior loops, and are therefore not consistent across multiple trials. 
    See the following issues for more info:
    - https://github.com/optuna/optuna/issues/372
    - https://github.com/optuna/optuna/issues/2199

2. (If dynamic categorical spaces would hypothetically be supported) Since the order of the subset is irrelevant, indexing each sampled variable with `f'column_{i}'` is inefficient from the optimization standpoint (yet required for the uniqueness of distribution names). A hacky and inefficient workaround would be to run a trial once, then use [study.add_trial](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html?highlight=enqueue_trial#optuna.study.Study.add_trial) to add all permutations of the sampled set to the study with the same objective value without re-evaluating the objective functions. 

Thank you and apologies if this has been discussed before.

**Comment by isole-01** [comment]
# Select a subset of columns using True/False for each feature
    selected_features = [trial.suggest_categorical(name, [True, False]) for name in columns]
    
    # Get the actual feature names that are selected
    selected_feature_names = [name for name, selected in zip(columns, selected_features) if selected]
    
    
    # penalize
    if Lowerbound < len(selected_feature_names) < Upperbound:
        return -1  
        
        
        
 you can use binary masking for your set and if the number of selected values doesn't work for you, you can penalize it with a high or low value
